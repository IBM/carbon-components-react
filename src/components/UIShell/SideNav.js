import cx from 'classnames';
import React from 'react';
import PropTypes from 'prop-types';
import isRequiredOneOf from '../../prop-types/isRequiredOneOf';
import SideNavHeader from './SideNavHeader';
import SideNavItems from './SideNavItems';
import SideNavFooter from './SideNavFooter';

// Variants
// Header -> with menu, without menu, with long title, with normal title
// Items
// Link with no icon, active/non-active
// Link with icon, active/non-active
// Menu with icon, non-collapsible, active/non-active child link
// Menu with icon, collapsible, active/non-active child link
// Menu without icon, non-collapsible, active/non-active child link
// Menu without icon, collapsible, active/non-active child link
// Sections
// Side nav with items that trigger overflow
// Side nav with typical items
// Side nav footer section (fixed?)
// Each item with a lot of text
// Text with RTL support?

const translations = {
  'carbon.sidenav.state.open': 'Close',
  'carbon.sidenav.state.closed': 'Open',
};

function translateById(id) {
  return translations[id];
}

export default class SideNav extends React.Component {
  static propTypes = {
    /**
     * Required props for accessibility label on the underlying menu
     */
    ...isRequiredOneOf({
      'aria-label': PropTypes.string,
      'aria-labelledby': PropTypes.string,
    }),

    /**
     * Optionally provide a custom class to apply to the underlying <li> node
     */
    className: PropTypes.string,

    /**
     * Provide a custom function for translating all message ids within this
     * component. This function will take in two arguments: the mesasge Id and the
     * state of the component. From this, you should return a string representing
     * the label you want displayed or read by screen readers.
     */
    translateById: PropTypes.func,
  };

  static defaultProps = {
    translateById,
  };

  state = {
    isExpanded: true,
  };

  handleExpand = () => {
    this.setState(state => ({ isExpanded: !state.isExpanded }));
  };

  render() {
    const {
      'aria-label': ariaLabel,
      'aria-labelledby': ariaLabelledBy,
      children,
      className: customClassName,
      translateById: t,
    } = this.props;
    const { isExpanded } = this.state;
    const accessibilityLabel = {
      'aria-label': ariaLabel,
      'aria-labelledby': ariaLabelledBy,
    };
    const assistiveText = isExpanded
      ? t('carbon.sidenav.state.open')
      : t('carbon.sidenav.state.closed');
    const className = cx({
      'bx--side-nav': true,
      'bx--side-nav--expanded': isExpanded,
      [customClassName]: !!customClassName,
    });

    return (
      <aside className={className}>
        <nav
          className="bx--side-nav__navigation"
          role="navigation"
          {...accessibilityLabel}>
          {children}
          <SideNavFooter
            assistiveText={assistiveText}
            isExpanded={isExpanded}
            onToggle={this.handleExpand}
          />
        </nav>
      </aside>
    );
  }
}
