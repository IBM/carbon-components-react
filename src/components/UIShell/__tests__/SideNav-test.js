/**
 * Copyright IBM Corp. 2016, 2018
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

import React from 'react';
import { mount } from 'enzyme';
import SideNav from '../SideNav';

describe('SideNav', () => {
  let mockProps;

  beforeEach(() => {
    mockProps = {
      'aria-label': 'Navigation',
      children: <h2>Navigation</h2>,
    };
  });

  it('should render', () => {
    const wrapper = mount(<SideNav {...mockProps} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('should toggle the menu expansion state when clicking on the footer', () => {
    const wrapper = mount(<SideNav {...mockProps} />);
    expect(wrapper.state('isExpanded')).toBe(false);
    wrapper.find('button').simulate('click');
    expect(wrapper.state('isExpanded')).toBe(true);
    expect(wrapper).toMatchSnapshot();
  });

  it('should trigger menu expansion state byDefault', () => {
    const wrapper = mount(<SideNav {...mockProps} defaultExpanded />);
    expect(wrapper.state('isExpanded')).toBe(true);
  });

  it('should add focus class when nav items are focused', () => {
    const wrapper = mount(<SideNav {...mockProps} />);
    wrapper.simulate('focus');
    expect(wrapper.state('isFocused')).toBe(true);
  });
});
