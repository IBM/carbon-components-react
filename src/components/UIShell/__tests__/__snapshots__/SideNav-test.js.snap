// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SideNav should render 1`] = `
<SideNav
  aria-label="Navigation"
  translateById={[Function]}
>
  <aside
    className="bx--side-nav bx--side-nav--expanded"
  >
    <nav
      aria-label="Navigation"
      className="bx--side-nav__navigation"
      role="navigation"
    >
      <h2>
        Navigation
      </h2>
      <SideNavFooter
        assistiveText="Close"
        isExpanded={true}
        onToggle={[Function]}
      >
        <footer
          className="bx--side-nav__footer"
        >
          <button
            className="bx--side-nav__toggle"
            onClick={[Function]}
            title="Close"
            type="button"
          >
            <div
              className="bx--side-nav__icon"
            >
              <Close>
                <svg
                  aria-hidden="true"
                  height="20"
                  viewBox="0 0 32 32"
                  width="20"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M17.414 16L24 9.414 22.586 8 16 14.586 9.414 8 8 9.414 14.586 16 8 22.586 9.414 24 16 17.414 22.586 24 24 22.586 17.414 16z"
                  />
                </svg>
              </Close>
            </div>
            <span
              className="bx--assistive-text"
            >
              Close
            </span>
          </button>
        </footer>
      </SideNavFooter>
    </nav>
  </aside>
</SideNav>
`;

exports[`SideNav should toggle the menu expansion state when clicking on the footer 1`] = `
<SideNav
  aria-label="Navigation"
  translateById={[Function]}
>
  <aside
    className="bx--side-nav"
  >
    <nav
      aria-label="Navigation"
      className="bx--side-nav__navigation"
      role="navigation"
    >
      <h2>
        Navigation
      </h2>
      <SideNavFooter
        assistiveText="Open"
        isExpanded={false}
        onToggle={[Function]}
      >
        <footer
          className="bx--side-nav__footer"
        >
          <button
            className="bx--side-nav__toggle"
            onClick={[Function]}
            title="Open"
            type="button"
          >
            <div
              className="bx--side-nav__icon"
            >
              <ChevronRight>
                <svg
                  aria-hidden="true"
                  height="20"
                  viewBox="0 0 32 32"
                  width="20"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M22 16L12 26l-1.414-1.414L19.172 16l-8.586-8.586L12 6l10 10z"
                  />
                </svg>
              </ChevronRight>
            </div>
            <span
              className="bx--assistive-text"
            >
              Open
            </span>
          </button>
        </footer>
      </SideNavFooter>
    </nav>
  </aside>
</SideNav>
`;
